### PR TITLE
Topic/scwin3.7 fix conflicts

### PR DESCRIPTION
**Pull request by telephon**
Let's see if this passes the tests.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,15 +15,41 @@ enable_testing()
 
 include (cmake_modules/FinalFile.cmake)
 
-if (NOT CMAKE_BUILD_TYPE)
-	message(STATUS "Build type defaulting to \"RelWithDebInfo\"")
-	set(CMAKE_BUILD_TYPE "RelWithDebInfo")
+if(CMAKE_GENERATOR MATCHES "Visual Studio 12 2013")
+    set(CMAKE_CONFIGURATION_TYPES Debug Release)
```
Comment (timblechmann) [inline CMakeLists.txt]: what's the reason for disabling the other configuration types? please remove

File: CMakeLists.txt [hunk 2]
```diff
@@ -15,15 +15,41 @@ enable_testing()
...
 
 if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
 	message(WARNING "WARNING: IN-PLACE BUILDS ARE NOT RECOMMENDED - PLEASE USE A BUILD DIRECTORY")
 endif()
 
+##############################################################################################
+# Win - Create target architecture Variable and set install path to SC default (cached system
+# default location throws error)
+if(WIN32)
+    if(CMAKE_GENERATOR MATCHES "64" OR "$ENV{MSYSTEM}" MATCHES "64")
```
Comment (timblechmann) [inline CMakeLists.txt]: use: CMAKE_CL_64

File: CMakeLists.txt [hunk 3]
```diff
@@ -240,8 +272,10 @@ if (CMAKE_COMPILER_IS_INTEL AND NOT WIN32)
 	add_definitions(-diag-disable 279) # controlling expression is constant
 endif()
 
-if(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
+if("x${CMAKE_CXX_COMPILER_ID}" STREQUAL "xMSVC")
```
Comment (timblechmann) [inline CMakeLists.txt]: no need for the "x"

File: CMakeLists.txt [hunk 4]
```diff
@@ -240,8 +272,10 @@ if (CMAKE_COMPILER_IS_INTEL AND NOT WIN32)
...
   foreach(flag CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
+    set(${flag} "${${flag}} /wd4520") # multiple default constructors (boost::flyweight)
+	set(${flag} "${${flag}} /wd4065") # switch statement contains 'default' but no 'case' labels
```
Comment (timblechmann) [inline CMakeLists.txt]: please fix the formatting

File: CMakeLists.txt [hunk 5]
```diff
@@ -361,7 +404,10 @@ else()
 	)
 endif()
 
-if(NOT APPLE)
+if(CMAKE_GENERATOR MATCHES "Visual Studio 12 2013")
+	install(DIRECTORY sounds   CONFIGURATIONS Debug;Release DESTINATION "${auxresourcesdir}")
```
Comment (timblechmann) [inline CMakeLists.txt]: remove: please do not disable the default configurations

File: CMakeLists.txt [hunk 6]
```diff
@@ -395,7 +441,29 @@ else()
...
+
+    install(FILES ${AdditionalInstallFiles} DESTINATION "${CMAKE_INSTALL_PREFIX}/${auxresourcesdir}")
+
+    if(MSYS)
+        SET(QT_PLATFORM_PLUGIN_PREFIX ${Qt5Core_DIR}/../../../share/qt5/plugins/platforms CACHE STRING "Location of qt platform plugin for windows. Override when not using MinGW-cmake!")
+    else()
+        SET(QT_PLATFORM_PLUGIN_PREFIX ${Qt5Core_DIR}/../../../plugins/platforms CACHE STRING "Location of qt platform plugin for windows. Override if you have multiple paths in your cmake prefix path")
+    endif()
+
+    install(FILES "${QT_PLATFORM_PLUGIN_PREFIX}/qwindows.dll"
```
Comment (timblechmann) [inline CMakeLists.txt]: is it possible to run the build without installing? it doesn't look like, so you might want to copy qt's platform plugins via a post-build step

File: QtCollider/style/ProxyStyle.cpp [hunk 7]
```diff
@@ -19,9 +19,9 @@ static bool AlwaysShowScrollbars() {
 #elif Q_OS_X11
   return !QcApplication::SystemHasMouseWheel();
 
-#elif Q_OS_WIN
-  return !QcApplication::SystemHasMouseWheel();
-
+// #elif Q_OS_WIN
+//  return !QcApplication::SystemHasMouseWheel();
+//
```
Comment (timblechmann) [inline QtCollider/style/ProxyStyle.cpp]: this is not a comment: either delete or fix.

File: editors/sc-ide/CMakeLists.txt [hunk 8]
```diff
@@ -371,7 +373,15 @@ if(APPLE)
                 #fixup_qt5_executable(\"${CMAKE_INSTALL_PREFIX}/SuperCollider/SuperCollider.app\")
                 " COMPONENT Runtime)
 
-
+elseif(CMAKE_GENERATOR MATCHES "Visual Studio 12 2013")
+    set_property(TARGET SuperCollider PROPERTY RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/editors/sc-ide")
+	install(TARGETS SuperCollider
+            CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline editors/sc-ide/CMakeLists.txt]: please don't disable other configurations

File: editors/sc-ide/CMakeLists.txt [hunk 9]
```diff
@@ -371,7 +373,15 @@ if(APPLE)
...
+			DESTINATION "SuperCollider"
+			PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)    
+    install(FILES ${translations}
+            CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline editors/sc-ide/CMakeLists.txt]: i'm skipping comments about this now

File: editors/sc-ide/core/main.cpp [hunk 10]
```diff
@@ -56,7 +56,7 @@ int main( int argc, char *argv[] )
     SingleInstanceGuard guard;
     if (guard.tryConnect(arguments))
         return 0;
-
+
```
Comment (timblechmann) [inline editors/sc-ide/core/main.cpp]: no trailing spaces please

File: editors/sc-ide/core/main.cpp [hunk 11]
```diff
@@ -128,72 +128,56 @@ int main( int argc, char *argv[] )
...
-
-                stream << QString("open");
-                stream << canonicalArguments;
-                if (!socket->waitForBytesWritten(300))
-                    qWarning("SingleInstanceGuard: writing data to another IDE instance timed out");
-
-                return true;
-            }
+        QTcpSocket *socket = new QTcpSocket(this);
+		socket->connectToHost(QHostAddress(QHostAddress::LocalHost), SingleInstanceGuard::Port);
```
Comment (timblechmann) [inline editors/sc-ide/core/main.cpp]: don't mix tabs and spaces

File: editors/sc-ide/core/main.cpp [hunk 12]
```diff
@@ -128,72 +128,56 @@ int main( int argc, char *argv[] )
...
+            canonicalArguments << info.canonicalFilePath();
         }
-    }
 
-    mIpcServer = new QLocalServer(this);
-    for (int socketID = 0; socketID != maxNumberOfInstances; ++socketID) {
-        QString serverName = QString("SuperColliderIDE_Singleton_%1").arg(socketID);
+        if (socket->waitForConnected(200)) {
+			ScIpcChannel *ipcChannel = new ScIpcChannel(socket, QString("SingleInstanceGuard"), this);
+			ipcChannel->write(QString("open"), canonicalArguments);
```
Comment (timblechmann) [inline editors/sc-ide/core/main.cpp]: QStringLiteral

File: editors/sc-ide/core/main.cpp [hunk 13]
```diff
@@ -128,72 +128,56 @@ int main( int argc, char *argv[] )
...
-        if (listening) {
-            connect(mIpcServer, SIGNAL(newConnection()), this, SLOT(onNewIpcConnection()));
-            return false;
+			return true;
         }
     }
-    return false;
+
+    mIpcServer = new QTcpServer(this);
+	bool listening = mIpcServer->listen(QHostAddress(QHostAddress::LocalHost), SingleInstanceGuard::Port);
```
Comment (timblechmann) [inline editors/sc-ide/core/main.cpp]: i'm not adding any comments about mixing tabs and spaces anymore. please review all the code in an ide which can visualise them

File: editors/sc-ide/core/sc_introspection.cpp [hunk 14]
```diff
@@ -72,7 +72,7 @@ bool Introspection::parse(const QString & yamlString )
 
     clear();
 
-    //qDebug("parsing introspection...");
+    qDebug("parsing introspection...");
```
Comment (timblechmann) [inline editors/sc-ide/core/sc_introspection.cpp]: please do not remove code by commenting out. there is absolutely no point in commenting out this, btw

File: editors/sc-ide/core/sc_process.cpp [hunk 15]
```diff
@@ -39,12 +39,18 @@ namespace ScIDE {
...
     mIpcServerName("SCIde_" + QString::number(QCoreApplication::applicationPid())),
     mTerminationRequested(false),
     mCompiled(false)
 {
+
+	mIpcServer->listen(QHostAddress(QHostAddress::LocalHost), ScIpcChannel::Port);
+
+	// this is initialized when connecting.
+	mIpcChannel = nullptr;
+	mIpcSocket = nullptr;
```
Comment (timblechmann) [inline editors/sc-ide/core/sc_process.cpp]: please initialise the members either in the class (we are finally using c++11) or as in the code you've just removed.

File: editors/sc-ide/core/sc_process.cpp [hunk 16]
```diff
@@ -59,6 +65,10 @@ ScProcess::ScProcess( Settings::Manager * settings, QObject * parent ):
     connect(this, SIGNAL(stateChanged(QProcess::ProcessState)), this, SLOT(onProcessStateChanged(QProcess::ProcessState)));
 }
 
+ScProcess::~ScProcess() {
+	finalizeConnection();
```
Comment (timblechmann) [inline editors/sc-ide/core/sc_process.cpp]: { on new line please

File: editors/sc-ide/core/sc_process.cpp [hunk 17]
```diff
@@ -308,101 +328,83 @@ void ScProcess::postQuitNotification()
...
 }
 
 void ScProcess::onStart()
 {
     if(!mIpcServer->isListening()) // avoid a warning on stderr
-        mIpcServer->listen(mIpcServerName);
+		mIpcServer->listen(QHostAddress(QHostAddress::LocalHost), ScIpcChannel::Port);
 
-    QString command = QString("ScIDE.connect(\"%1\")").arg(mIpcServerName);
+	QString command = QString("ScIDE.connect(\"%1\")").arg(mIpcServerName);
```
Comment (timblechmann) [inline editors/sc-ide/core/sc_process.cpp]: ???

File: editors/sc-ide/core/sc_process.cpp [hunk 18]
```diff
@@ -308,101 +328,83 @@ void ScProcess::postQuitNotification()
...
+	qDebug() << message;
+	//scPost(message);
+}
+
+
+void ScProcess::onIpcMessage(const QString & selector, const QVariantList & data)
+{
+	static QString introspectionSelector("introspection");
+	static QString classLibraryRecompiledSelector("classLibraryRecompiled");
+	static QString requestCurrentPathSelector("requestCurrentPath");
```
Comment (timblechmann) [inline editors/sc-ide/core/sc_process.cpp]: we are using qt5, so QStringLiteral is the way to go

File: editors/sc-ide/core/settings/theme.cpp [hunk 19]
```diff
@@ -247,8 +247,8 @@ void Theme::setFormat(const QString & key, const QTextCharFormat & newFormat)
                     QColor(Qt::transparent): newFormat.background().color();
 
     if (i == mFormats.end()) {
-        qDebug() << __func__ << "Failed to find key " << key;
-        return;
+        qDebug() <<  __FUNCTION__ << "Theme::setFormat" << "Failed to find key " << key;
+		return;
     }
```
Comment (timblechmann) [inline editors/sc-ide/core/settings/theme.cpp]: ???

File: editors/sc-ide/primitives/sc_ipc_channel.hpp [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+#include <QByteArray>
+#include <QIODevice>
+#include <QBuffer>
+#include <QTcpSocket>
+#include <QSharedMemory>
+#include <QThread>
+#include <QMutex>
+
+struct IIpcHandler {
+	virtual void onIpcLog(const QString &message) = 0;
```
Comment (timblechmann) [inline editors/sc-ide/primitives/sc_ipc_channel.hpp]: Qt has the convention of passing by value,  not by reference

File: lang/CMakeLists.txt [hunk 21]
```diff
@@ -325,6 +324,12 @@ if(APPLE)
 	install(TARGETS sclang
 			DESTINATION "${scappbindir}"
 			PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)
+elseif(CMAKE_GENERATOR MATCHES "Visual Studio 12 2013")
+    set_property(TARGET  sclang PROPERTY RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lang")
+    install(TARGETS sclang
+        CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline lang/CMakeLists.txt]: ...

File: lang/LangPrimSource/PyrSched.cpp [hunk 22]
```diff
@@ -1,22 +1,22 @@
...
+ 
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
+ */
```
Comment (timblechmann) [inline lang/LangPrimSource/PyrSched.cpp]: please revert the whitespace change

File: lang/LangPrimSource/PyrSched.cpp [hunk 23]
```diff
@@ -50,7 +54,7 @@ void runAwakeMessage(VMGlobals *g);
 
 // heaps use an integer timestamp to ensure stable heap order
 struct PyrHeap:
-	PyrObjectHdr
+PyrObjectHdr
```
Comment (timblechmann) [inline lang/LangPrimSource/PyrSched.cpp]: please revert

File: lang/LangPrimSource/PyrSched.cpp [hunk 24]
```diff
@@ -67,16 +71,16 @@ bool addheap(VMGlobals *g, PyrObject *heapArg, double schedtime, PyrSlot *task)
...
 		return false;
 	assert(heap->size);
-
-// 	post("->addheap\n");
-// 	dumpheap(heapArg);
-
+	
+	// 	post("->addheap\n");
+	// 	dumpheap(heapArg);
+
```
Comment (timblechmann) [inline lang/LangPrimSource/PyrSched.cpp]: ?

File: lang/LangPrimSource/PyrSched.cpp [hunk 25]
```diff
@@ -67,16 +71,16 @@ bool addheap(VMGlobals *g, PyrObject *heapArg, double schedtime, PyrSlot *task)
...
+	// 	post("->addheap\n");
+	// 	dumpheap(heapArg);
+	
 	/* parent and sibling in the heap, not in the task hierarchy */
 	int mom = heap->size - 1;
 	PyrSlot * pme = heap->slots + mom;
 	int stabilityCount = slotRawInt(&heap->count);
 	SetRaw(&heap->count, stabilityCount + 1);
-
+
```
Comment (timblechmann) [inline lang/LangPrimSource/PyrSched.cpp]: i'm skipping comments about whitespaces in this file. please fix

File: lang/LangSource/SC_TerminalClient.cpp [hunk 26]
```diff
@@ -298,8 +298,8 @@ static PyrSymbol * resolveMethodSymbol(bool silent)
 {
 	if (silent)
 		return s_interpretCmdLine;
-	else
-		return s_interpretPrintCmdLine;
+	
+	return s_interpretPrintCmdLine;
```
Comment (timblechmann) [inline lang/LangSource/SC_TerminalClient.cpp]: i know this is no functional change, but it doesn't make the code exactly readable

File: server/plugins/CMakeLists.txt [hunk 27]
```diff
@@ -257,9 +257,18 @@ elseif (APPLE)
 endif()
 
 if (WIN32)
-	install(TARGETS ${plugins} ${supernova_plugins}
+    if(CMAKE_GENERATOR MATCHES "Visual Studio 12 2013")
+        set_property(TARGET  ${plugins} ${supernova_plugins} PROPERTY LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/server/plugins")
+    install(TARGETS ${plugins} ${supernova_plugins}
+        CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline server/plugins/CMakeLists.txt]: .

File: server/scsynth/CMakeLists.txt [hunk 28]
```diff
@@ -187,7 +187,10 @@ else()
 endif()
 
 if (WIN32)
-	target_link_libraries(libscsynth wsock32 ws2_32 winmm)
+	target_link_libraries(libscsynth wsock32 ws2_32 winmm kernel32 user32
+		gdi32 winspool shell32 ole32 oleaut32 uuid comdlg32 advapi32 tlsf
+		boost_thread boost_system)
```
Comment (timblechmann) [inline server/scsynth/CMakeLists.txt]: whty are tlsf/boost.thread/system only needed on windows? this looks fishy

**Comment by timblechmann** [comment]
ok, after reviewing this PR, i'm having several comments:

* it hardcodes behaviour for msvc2013, while msvc2015 is about to be released. in many cases, this is not necessary or should be addressed in a more general way.

* it messes the whitespaces. PyrSched.cpp is completely messed up and the whitespace changes make it hard to review the actual functional change.

* apparently local sockets are replaced by tcp sockets. i'm quite surprised that this is the way to go and would like to hear the reason for this. 

* some changes seem to be semantically wrong

-----------------

this PR is "all-or-nothing": some changes are correct, some aren't. i'd ask the submitters to split it into independent pull requests, which can be reviewed and merged individually. if things are simply hacked together they might work, but will cause maintenance problems in the long term. i'm more than happy to review individual parts.

**Comment by bagong** [comment]
Thanks for your review, Tim. I'll be working on this again from the 16th onwards...

File: server/scsynth/SC_Lib_Cintf.cpp [hunk 29]
```diff
@@ -46,7 +46,7 @@
 
 // Plugin directory in resource directory
 #if defined(_WIN32) && defined(_DEBUG)
-# define SC_PLUGIN_DIR_NAME "plugins_debug"
+# define SC_PLUGIN_DIR_NAME "plugins"
 #else
 # define SC_PLUGIN_DIR_NAME "plugins"
 #endif
```
Comment (timblechmann) [inline server/scsynth/SC_Lib_Cintf.cpp]: there is no point for the ifdef anymore, no?

**Comment by sofakid** [comment]
Re: apparently local sockets are replaced by tcp sockets

It was frustrating, and I tried a bunch of stuff before resorting to TCP. 

QLocalSockets are implemented as Named Pipes on windows. Poorly. After 4K had been written to the pipe, it would block writing, effectively freezing the IDE.

As well, the other side would unblock immediately, when there was barely any bytes available. So it needed to wait until all the data was available. It would not be available because the sender was blocked after the first 4K. 

So we'd have to make some kind of buffer.. TCP gives us this at the expense of a port number.

Remember we can't use QT's overloaded << until all the data is there.. 

I had a solution that used shared memory when the messages were large, but on Linux, QT uses sysV shared memory rather than mmap(), with an arbitrary limit imposed on the size of the segment... Leading to needing to be root to increase it to fit the 7meg introspection string. Which sucked.

It was frustrating, and I had to make a lot of decisions that I internally disagreed with.

**Comment by sofakid** [comment]
Re: some changes seem to be semantically wrong

I don't doubt that. The code for the SingleInstanceGuard, I'm not sure about that. But the old code didn't make sense in the context of the new IPC design.

**Comment by sofakid** [comment]
Re: msvc2013 with 2015 out now.

Let's first target 2013. 2015 is a whole new thing. This PR is already huge.

**Comment by timblechmann** [comment]
@sofakid as mentioned: being a huge pull request is the reason, why i'm strongly against merging it. please split it to functional units, which can be reviewed and merged independently.

* problem is not 2013 or 2015: problem is hardcoding it!

* QLocalSockets: if this is a problem on windows only, it should be #ifdeffed out for windows. btw, if it is really a bug with QLocalSocket on windows, please submit a bug report to qt and mention the bug id in the sources ....

* tlsf is only needed on windows? this is clearly semantically wrong.

--

so please split the PR into separate units and submit them for review one by one.

**Comment by sofakid** [comment]
I don't have time for that.

**Comment by sofakid** [comment]
nor motivation.

**Comment by timblechmann** [comment]
@sofakid, someone has to maintain this code (read: devote time) in future and some part of this pull request is valuable, some isn't (or: some is unnecessary, messy or simply wrong).

fwiw, i will vote against merging this as it is, due to the issues i mentioned before.

**Comment by jamshark70** [comment]
A couple of cents:

- The IPC changes are absolutely essential. Without them, there is no Windows version anymore.

- Lucas took on the job of rescuing the Windows version when no one else would, and he did it basically single-handedly. I think it would be a real shame if the community's response to this were, "OK, Lucas, now you are single-handedly responsible for doing hours more work to make it ready for merge" rather than "How can we as a community work together to make it ready for merge?"

- Splitting the PR into functional units depends on some grunt work with code editors (lots of people could help with that), but that also depends on defining useful functional units. I don't feel I can help with that because all of this is way beyond my meager C++ knowledge. Tim, I can completely understand why you would not be interested in doing the grunt work, but perhaps you could offer some guidelines on pull requests you would approve?

**Comment by jamshark70** [comment]
TL;DR the current standoff is not productive:

- Tim: "This is not good enough to merge yet" (which seems objectively true).
- Lucas: "I've done enough" (which is understandable).

So what do we do to move beyond this?

**Comment by telephon** [comment]
This is really an important large step, and Lucas, you've really done a good job. I would like to help, but without a windows infrastructure I can't. What should we do to get enough people together who could actually maintain this code?

We should always keep this branch mergeable.

Smaller parts would help, but this is a lot of interdependent stuff, so it may be non-trivial to separate.

**Comment by timblechmann** [comment]
@telephon things should be fairly easy to separate: just have a look at the diff and get the good parts out. no need to have a windows infrastructure for it.

**Comment by telephon** [comment]
@timblechmann thank you for the confidence in my abilities. I had another close look, but I really shouldn't  intervene in issues I don't understand.

**Comment by timblechmann** [comment]
this is mostly obsoleted by #1692 ... for the IDE/localsocket problem, i don't think it is the right way to go ... especially since qt can be configured to implement local sockets via tcp (compare QT_LOCALSOCKET_TCP) ... if this is really a problem, it should (a) be reported to the qt developers and (b) the winrt/wince codepaths in qtbase/src/network/socket/socket.pri should be enabled for win32 as well.

**Comment by crucialfelix** [comment]
status ? Did the other windows fix (already merged) contain the main body of this work ?

**Comment by bagong** [comment]
Yes, this is obsolete... The controversial part, the "IPC-fix" is separated out in topic branch (topic/ipc_qtcp) the rest is in merged branch. This can be closed. Right, Julian?

**Comment by bagong** [comment]
@timblechmann, btw, this very IPC-bug, as well the breakage of console mode/readline and midi have been with the Windows version since you introduced boost-asio back then, and thereby broke the Windows version. It's been there through all variations of build systems from Qt 4.8 up right into 5.5.1 Yes, it is also you who brought Windows back during the work on HID, but you didn't finish the job...

**Comment by timblechmann** [comment]
@bagong what's your point?

**Comment by bagong** [comment]
@timblechmann Just for the record

**Comment by timblechmann** [comment]
@bagong just for the record

* boost.asio and qt/ipc are completely separate systems and i've never supported the win32 port. 
* `topic/ipc_qtcp` is going to introduce at least one bug, and using tcp instead of local sockets isn't the right way to fix it. however my recommendation to analyse the issue and file an upstream bug report has been ignored.
* i probably fixed an order of magnitude more bugs in sc than i've introduced. usually without personally blaming the people, who introduced them. 
* if i hadn't spent a significant amount of time on the language, sc would still be 32bit-only.

... just for the record ...

**Comment by bagong** [comment]
@timblechmann, I am not trying to go into confrontation. My point is: 
- since boost.asio was introduced, the Windows build was broken. The important thing from my point of view is that a commit was made that breaks platform support, without letting others know. There is an important difference between giving up support for a platform as result of a consensual community decision, and single handedly deciding to give up a platform for the benefit of others.
- the symptoms of the "IPC-bug" have been there since you recovered the Windows build from it's out-time during work on the new HID support. As I remember better now, this work was done using Qt 5.2, so unfortunately there is no base to compare a pre boost.asio sc3.7win without the symptoms, with a post-HID one, as back then Qt4 was used.
- I didn't even say you "introduced" a bug, I just said it is a left-over of your work on Windows support. This is a useful pointer if anybody tries to work on the problems. What I stand by though, is that part of your commits do not take Windows into consideration. I my view this is unacceptable as long as there is no *community* decision to stop supporting Windows. In effect your disregard of the Windows platform has broken it, and you did not enquire whether this is okay.
- "usually without personally blaming the people, who introduced them". Call it blaming, call it making things transparent, that's your choice. I think it is a lot better than calling people's work crap, who tried hard within the limits of their capabilities, to get out of a situation that you have created.
- I am not questioning your achievements in the least. You know that I have great respect for your capabilities. All I am asking is that you consult others before deciding to break a platform.
- if topic/ipc_qtcp is introducing at least one bug, you would do us a great favor by naming it. As it looks right now there will be no alternative to using that branch *for the Windows version of SC* until a more acceptable fix is made.

**Comment by timblechmann** [comment]
i've mentioned the bug more more than once. was probably the last review you got from me

**Comment by bagong** [comment]
I would regret that...

**Comment by crucialfelix** [comment]
Let me just say that I woke up this morning thinking about when I can announce the finished 3.7 release and I was thinking of who to give special thanks to on behalf of the sc community. both of you guys were definitely on that list.

so we all thank you two. even the arguing here might be useful since we do improve our development process through these.

**Comment by telephon** [comment]
it's hard not to sound pathetic, so simply: you two have deserved more than a few beers!

**Comment by crucialfelix** [comment]
but tim doesn't like beer.

**Comment by bagong** [comment]
I'll drink his ;)
